task6-gpio-and-timer/module.c

```diff
@@ -0,0 +1,142 @@
+#include <linux/init.h>         // Macros used to mark up functions e.g., __init __exit
+#include <linux/module.h>       // Core header for loading LKMs into the kernel
+#include <linux/kernel.h>       // Contains types, macros, functions for the kernel
+#include <linux/moduleparam.h>
+#include <linux/proc_fs.h>
+#include <linux/uaccess.h>
+#include <linux/timer.h>
+#include <linux/gpio.h>
+#include <linux/delay.h>		//usleep(), msleep()
+#include <linux/jiffies.h>
+
+
+MODULE_LICENSE("GPL");                  ///< The license type -- this affects runtime behavior
+MODULE_AUTHOR("Yevhen Kolesnyk");       ///< The author -- visible when you use modinfo
+MODULE_DESCRIPTION("A simple device driver - Kernel Timer with GPIO");  ///< The description -- see modinfo
+MODULE_VERSION("0.1");                  ///< The version of the module
+
+#define TIMEOUT         1000 //milliseconds
+#define PROC_BUFFER_SIZE 100
+#define GPIO_4 (4) //LED is connected to this GPIO
+
+static int led_state; // 0 - disabled, 1 - enabled
+
+static struct timer_list  etx_timer;
+static unsigned int count_timer;
+
+static char procfs_buffer[PROC_BUFFER_SIZE] = {0};
+static size_t procfs_buffer_size;
+
+static struct proc_dir_entry *proc_file;
+static struct proc_dir_entry *proc_folder;
+
+#define PROC_FILE_NAME "LED_blink"
+#define PROC_DIR_NAME "GPIO_4"
+
+
+static ssize_t gpio_read(struct file *File, char __user *buffer, size_t count, loff_t *offset)
+{
+	ssize_t to_copy, not_copied, delta;
+
+	if (procfs_buffer_size == 0)
+		return 0;
+
+	to_copy = min(count, procfs_buffer_size);
+
+	not_copied = copy_to_user(buffer, procfs_buffer, to_copy);
+
+	delta = to_copy - not_copied;
+	procfs_buffer_size -= delta;
```

Comment:
This comment is out of scope of the current task, just to let you know:

Here is a potentially unsafe operation. The `procfs_buffer_size` can be suddenly updated in the `timer_callback` function below, which will lead to race conditions and unexpected behavior.